File: ckgame-ts/src/test/GameSimulation.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { GameSimulation } from '../game/GameSimulation.js';

test('按日推进更新日期与计数', () => {
  const sim = new GameSimulation();
  const start = sim.world.date;
  sim.runDays(70);
  assert.equal((sim.world as any).tick, 70);
  assert.ok(sim.world.date.daysUntil(start) < 0, '日期应已前进');
  const elapsedMonths =
    (sim.world.date.year() - start.year()) * 12 + (sim.world.date.month() - start.month());
  assert.ok(elapsedMonths >= 2);
});

test('一个季度的模拟保持稳定', () => {
  const sim = new GameSimulation();
  sim.runDays(92); // 覆盖至少 3 次月度结算
  const rulers = sim.world.rulers();
  assert.ok(rulers.length >= 1);
  for (const r of rulers) {
    assert.ok(sim.world.character(r.id), `统治者 ${r.id} 应仍存在`);
  }
});

test('月初触发月度结算', () => {
  const sim = new GameSimulation();
  const startMonth = sim.world.date.month();
  for (let i = 0; i < 40; i++) {
    if (sim.world.date.day() === 1 && (sim.world as any).tick > 0) break;
    sim.tickDay();
  }
  assert.equal(sim.world.date.day(), 1);
  assert.notEqual(sim.world.date.month(), startMonth);
});

File: ckgame-java/src/test/java/com/ckgame/game/GameSimulationTest.java
package com.ckgame.game;

import com.ckgame.world.Character;
import com.ckgame.world.World;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/** 主循环稳定性测试：按日/按月推进不崩溃，状态保持一致。 */
class GameSimulationTest {

    @Test
    void runDaysAdvancesDateAndTick() {
        GameSimulation sim = new GameSimulation();
        World w = sim.world;
        int startYear = w.date.year();
        int startMonth = w.date.month();
        sim.runDays(70);
        assertEquals(70, w.tick);
        int elapsedMonths = (w.date.year() - startYear) * 12 + (w.date.month() - startMonth);
        assertTrue(elapsedMonths >= 2, "70 天应跨过至少两个月度结算");
    }

    @Test
    void quarterOfStableSimulation() {
        GameSimulation sim = new GameSimulation();
        sim.runDays(92);
        List<Character> rulers = sim.world.rulers();
        assertTrue(rulers.size() >= 1);
        for (Character r : rulers) {
            assertNotNull(sim.world.character(r.id));
        }
    }

    @Test
    void monthStartTriggersMonthlySettlement() {
        GameSimulation sim = new GameSimulation();
        World w = sim.world;
        int startMonth = w.date.month();
        for (int i = 0; i < 40; i++) {
            if (w.date.day() == 1 && w.tick > 0) break;
            sim.tickDay();
        }
        assertEquals(1, w.date.day());
        assertNotEquals(startMonth, w.date.month());
    }
}
